### PR TITLE
Closes #2818 - Supporting custom SSH URL's when hosting behind a proxy

### LINKED ISSUE #2818: Feature request: Custom SSH URL
## Issue
**Impacted version**: 4.35.3

**Deployment mode**: Tomcat 9 + PostgreSQL + Apache reverse proxy

**Problem description**:

螢幕擷取畫面 2021-07-11 174534

I setted port forward, The SSH URL displayed on the Web UI is not what I need :-(

I hope the SSH URL is configurable (Can even remove the port).

Thx. :-)

### PR DESCRIPTION
**Pull request by keywordsalad**
I've made changes to support hosting GitBucket behind a proxy. The public SSH host and port can now be configured independently of the locally-bound SSH host and port. When the public port is 22, the SSH url's leave off "ssh://" and the port number: they look like what everyone's used to on GitHub.

### Before submitting a pull-request to GitBucket I have first:

- [x] read the [contribution guidelines](https://github.com/gitbucket/gitbucket/blob/master/.github/CONTRIBUTING.md)
- [x] rebased my branch over master
- [x] verified that project is compiling
- [x] verified that tests are passing
- [x] squashed my commits as appropriate *(keep several commits if it is relevant to understand the PR)*
- [x] [marked as closed using commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) all issue ID that this PR should correct

### CODE REVIEW

File: src/main/scala/gitbucket/core/service/SystemSettingsService.scala [hunk 1]
```diff
@@ -112,9 +123,20 @@ trait SystemSettingsService {
...
+        enabled = getValue(props, SshEnabled, false),
+        bindAddress = {
+          // try the new-style configuration first
+          getOptionValue[String](props, SshBindAddressHost, None)
+            .map(h => SshAddress(h, getValue(props, SshBindAddressPort, DefaultSshPort), GenericSshUser))
+            .orElse(
+              // otherwise try to get old-style configuration
+              getOptionValue[String](props, SshHost, None)
+                .map(_.trim)
+                .map(h => SshAddress(h, getValue(props, SshPort, DefaultSshPort), GenericSshUser))
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/service/SystemSettingsService.scala:135]: Offering backword compatibility is nice!

File: src/main/scala/gitbucket/core/service/SystemSettingsService.scala [hunk 2]
```diff
@@ -296,14 +348,34 @@ object SystemSettingsService {
...
+    def getUrl: String =
+      if (isDefaultPort) {
+        s"${genericUser}@${host}"
+      } else {
+        s"${genericUser}@${host}:${port}"
+      }
+
+    def getUrl(owner: String, name: String): String =
+      if (isDefaultPort) {
+        s"${genericUser}@${host}:${owner}/${name}.git"
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/service/SystemSettingsService.scala:365]: Isn't `ssh://` necessary?

**Review by takezoe — COMMENTED** [review]
Looks like `bindAddress.host` is not necessary anymore. The actual configuration we need is:

- Local port
- Public host
- Public port (optional)

or it rather can be the following two items?

- Local port
- Display URL (`<host>` or `<host:port>`)

How about having only these three configuration items?

File: src/main/scala/gitbucket/core/service/SystemSettingsService.scala (hunk 2 above)
Comment (keywordsalad) [reply src/main/scala/gitbucket/core/service/SystemSettingsService.scala:365]: It isn’t when the port is 22! The only time it’s required is if the port is different.

Comment (takezoe) [reply src/main/scala/gitbucket/core/service/SystemSettingsService.scala:365]: Oh, I see. Thank you for the clarification.

File: src/main/scala/gitbucket/core/controller/SystemSettingsController.scala [hunk 3]
```diff
@@ -116,8 +127,8 @@ trait SystemSettingsControllerBase extends AccountManagementControllerBase {
       if (settings.ssh.enabled && settings.baseUrl.isEmpty) {
         Some("baseUrl" -> "Base URL is required if SSH access is enabled.")
       } else None,
-      if (settings.ssh.enabled && settings.ssh.sshHost.isEmpty) {
-        Some("sshHost" -> "SSH host is required if SSH access is enabled.")
+      if (settings.ssh.enabled && settings.ssh.bindAddress.isEmpty) {
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/controller/SystemSettingsController.scala:130]: Looks like `publicAddress` is also required if we want to enable SSH.

File: src/main/scala/gitbucket/core/ssh/SshServerListener.scala [hunk 4]
```diff
@@ -59,13 +59,14 @@ class SshServerListener extends ServletContextListener with SystemSettingsServic
 
   override def contextInitialized(sce: ServletContextEvent): Unit = {
     val settings = loadSystemSettings()
-    if (settings.sshAddress.isDefined && settings.baseUrl.isEmpty) {
+    if (settings.sshBindAddress.isDefined && settings.baseUrl.isEmpty) {
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/ssh/SshServerListener.scala:62]: `sshPublicAddress` is also required?

**Comment by takezoe** [comment]
I want to include this in the next release, so please allow me to merge this and modify it slightly. 🙇

File: src/main/scala/gitbucket/core/controller/SystemSettingsController.scala (hunk 3 above)
Comment (takezoe) [reply src/main/scala/gitbucket/core/controller/SystemSettingsController.scala:130]: Sorry, I overlooked that `sshBindAddress` is used as `sshPublicAddress` if `sshPublicAddress` is not given.

**Pull request merged**